src/typemap.c

```diff
@@ -19,9 +19,80 @@ extern "C" {
 // compute whether the specificity of this type is equivalent to Any in the sort order
 static int jl_is_any(jl_value_t *t1)
 {
-    return (t1 == (jl_value_t*)jl_any_type ||
-            (jl_is_typevar(t1) &&
-             ((jl_tvar_t*)t1)->ub == (jl_value_t*)jl_any_type));
+    while (jl_is_typevar(t1))
+        t1 = ((jl_tvar_t*)t1)->ub;
+    return t1 == (jl_value_t*)jl_any_type;
+}
+
+static jl_value_t *jl_type_extract_name(jl_value_t *t1, int abstract)
+{
+    if (jl_is_unionall(t1))
+        t1 = jl_unwrap_unionall(t1);
+    if (jl_is_vararg_type(t1)) {
+        return jl_type_extract_name(jl_tparam0(t1), abstract);
+    }
+    else if (jl_is_typevar(t1)) {
+        return jl_type_extract_name(((jl_tvar_t*)t1)->ub, abstract);
+    }
+    else if (jl_is_datatype(t1)) {
+        jl_datatype_t *dt = (jl_datatype_t*)t1;
+        if ((abstract || !dt->abstract) && !jl_is_kind(t1))
+            return (jl_value_t*)dt->name;
+        return NULL;
+    }
+    else if (jl_is_uniontype(t1)) {
+        jl_uniontype_t *u1 = (jl_uniontype_t*)t1;
+        jl_value_t *tn1 = jl_type_extract_name(u1->a, abstract);
+        jl_value_t *tn2 = jl_type_extract_name(u1->b, abstract);
+        if (tn1 == tn2)
+            return tn1;
+        // TODO: if abstract is true, instead find the nearest common ancester
+        return NULL;
+    }
+    return NULL;
+}
+
+// return true if the name extracted above is an over-approximation
+// (such that intersection also needs to consider the subtypes)
+static int jl_type_extract_name_imprecise(jl_value_t *t1, int abstract)
+{
+    if (jl_is_unionall(t1))
+        t1 = jl_unwrap_unionall(t1);
+    if (jl_is_vararg_type(t1)) {
+        return jl_type_extract_name_imprecise(jl_tparam0(t1), abstract);
+    }
+    else if (jl_is_typevar(t1)) {
+        return jl_type_extract_name_imprecise(((jl_tvar_t*)t1)->ub, 0);
+    }
+    else if (jl_is_datatype(t1)) {
+        jl_datatype_t *dt = (jl_datatype_t*)t1;
+        if ((abstract || !dt->abstract) && !jl_is_kind(t1))
+            return 0;
+        return 1;
```

Comment:
In this case we return `NULL` above, so in what sense is that imprecise? Or maybe a better question is, when does this function differ from `jl_type_extract_name() == NULL`?

Reply:
I suppose NULL is imprecise in the sense that we'll never consider asking this `jl_type_extract_name_imprecise` question if the result was NULL?